### PR TITLE
Fix the issue of Android add2app build fails on Android when assets are read-only

### PR DESCRIPTION
**Pull request by shihchanghsiungsonos**
## Description

With the latest dev build (Flutter 1.10.14-dev), the flutter:copyFlutterAssets task fails when the assets being copied are read-only. This can happen when a source control tool makes files read-only.

## Related Issues

#43686

## Tests

I added the following tests:

Update module_test.dart by add test cases to verify that read-only assets are copied to flutter_assets with user-readable and writable modes.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: *Replace with a link to your migration guide*

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Comment by christopherfujino** [comment]
@shihchanghsiungsonos looks like you have an analyzer error:

```
   info • Unnecessary new keyword • dev/devicelab/bin/tasks/module_test.dart:44:47 • unnecessary_new
```

**Comment by shihchanghsiungsonos** [comment]
> .

Thanks. Will fix it.

**Comment by jmagman** [comment]
@blasten Can you take a look?

**Review by blasten — APPROVED** [review]
Thanks @shihchanghsiungsonos! LGTM

**Comment by christopherfujino** [comment]
Looks like build tests are failing because of a dependency download timing out. I opened a discussion about it in the contributors' chat, and will try to see if I can get these green.

**Comment by dannyvalentesonos** [comment]
@blasten We just ran into another version of this read-only issue. So I had to dig into this again, and turns out that it's the `flutter` binary that copies the flutter_assets directory to the intermediatesDirectory as part of the compile task originally. So, I was able to now correct this in a better way by adding write permissions to files once the `compileTask` task completes.
We'll update this PR as soon as we can, and thus I'd recommend not merging this in just yet.
Thanks

**Comment by shihchanghsiungsonos** [comment]
PR request has been updated with Danny's latest solution. Thanks!

**Comment by dannyvalentesonos** [comment]
@blasten @jmagman We have updated this PR with the latest version of the changes. Now, after `flutter` runs, we add write permissions to files if needed, which fixes all downstream copies as well. We had several developers running into permission errors when running `./gradlew installDebug` and with this version of the fix, all issues are resolved.
Thanks

**Review by blasten — COMMENTED** [review]
The logic in flutter.gradle might need to work on Windows too. Have you experienced this issue on Windows?

**Comment by dannyvalentesonos** [comment]
> The logic in flutter.gradle might need to work on Windows too. Have you experienced this issue on Windows?

We do not build for Windows, so I don't know.

**Comment by dannyvalentesonos** [comment]
@blasten We've updated the code for Windows as well. Let us know what you think. Thanks.

File: dev/devicelab/bin/tasks/module_test.dart [hunk 1]
```diff
@@ -226,6 +252,30 @@ Future<void> main() async {
         );
       }
 
+      if (!Platform.isWindows) {
```
Comment (blasten) [inline dev/devicelab/bin/tasks/module_test.dart]: nit: this check can be removed.

File: dev/devicelab/bin/tasks/module_test.dart [hunk 2]
```diff
@@ -272,6 +322,31 @@ Future<void> main() async {
       ) {
         return TaskResult.failure("Release host APK doesn't contain metadata: flutterProjectType = module ");
       }
+
+      if (!Platform.isWindows) {
```
Comment (blasten) [inline dev/devicelab/bin/tasks/module_test.dart]: ditto

File: dev/devicelab/bin/tasks/module_test.dart (hunk 1 above)
Comment (shihchanghsiungsonos) [reply dev/devicelab/bin/tasks/module_test.dart]: The reason is I didn't set the file read only permission at line 54 above for Windows. So, just not checking r-w permission for Windows here and below.

Was thinking verify the file permission change for linux only should be enough.

File: packages/flutter_tools/gradle/flutter.gradle [hunk 3]
```diff
@@ -647,6 +647,15 @@ class FlutterPlugin implements Plugin<Project> {
                 extraGenSnapshotOptions extraGenSnapshotOptionsValue
                 splitDebugInfo splitDebugInfoValue
                 treeShakeIcons treeShakeIconsOptionsValue
+                doLast {
+                    project.exec {
+                        if (Os.isFamily(Os.FAMILY_WINDOWS)) {
+                            commandLine('cmd', '/c', 'attrib -r ${assetsDirectory}')
+                        } else {
+                            commandLine('chmod', '-R', 'u+w', assetsDirectory)
+			}
```
Comment (blasten) [inline packages/flutter_tools/gradle/flutter.gradle]: minor nit: indentation.

**Comment by blasten** [comment]
LGTM + minor nit

File: packages/flutter_tools/gradle/flutter.gradle [hunk 4]
```diff
@@ -647,6 +647,15 @@ class FlutterPlugin implements Plugin<Project> {
                 extraGenSnapshotOptions extraGenSnapshotOptionsValue
                 splitDebugInfo splitDebugInfoValue
                 treeShakeIcons treeShakeIconsOptionsValue
+                doLast {
+                    project.exec {
+                        if (Os.isFamily(Os.FAMILY_WINDOWS)) {
+                            commandLine('cmd', '/c', 'attrib -r ${assetsDirectory}')
```
Comment (dannyvalentesonos) [inline packages/flutter_tools/gradle/flutter.gradle]: I just realized that this does not run recursively. We need to run this instead:
`commandLine('cmd', '/c', 'attrib -r ${assetsDirectory}/* /s')`

File: packages/flutter_tools/gradle/flutter.gradle [hunk 5]
```diff
@@ -859,6 +868,10 @@ class FlutterTask extends BaseFlutterTask {
         return intermediateDir
     }
 
+    String getAssetsDirectory() {
+        return "${outputDirectory}/flutter_assets/"
```
Comment (dannyvalentesonos) [inline packages/flutter_tools/gradle/flutter.gradle]: we should remove the trailing / as it's not needed for chmod and with it we will have 2 slashes in the windows command above.

**Review by dannyvalentesonos — CHANGES_REQUESTED** [review]

**Comment by dannyvalentesonos** [comment]
@blasten Can you confirm everything is good with the latest changes? Can we get this merged? Thanks!

**Comment by blasten** [comment]
@dannyvalentesonos I saw a failure on Windows. I will take a look once the test completes.

**Comment by dannyvalentesonos** [comment]
Thanks @blasten. We found the issue and working on a fix.

**Comment by shihchanghsiungsonos** [comment]
> Thanks @blasten. We found the issue and working on a fix.

Fix has been committed. Thanks!

**Comment by blasten** [comment]
Thanks! LGTM

**Pull request merged**